src/qttools/datastructures/routines.py

```diff
@@ -67,21 +78,31 @@ def bd_matmul(
                     if out_range:
                         i_a, k_a = correct_out_range_index(i, k, num_blocks)
                         k_b, j_b = correct_out_range_index(k, j, num_blocks)
-                        tmp += a.blocks[i_a, k_a] @ b.blocks[k_b, j_b]
+                        partsum += a.blocks[i_a, k_a] @ b.blocks[k_b, j_b]
                     else:
-                        tmp += a.blocks[i, k] @ b.blocks[k, j]
-            out.blocks[i, j] = tmp
+                        partsum += a.blocks[i, k] @ b.blocks[k, j]
+
+            if out_block_coo:
+                if partsum.ndim <= 2:
+                    out[i][j] = sparse.coo_matrix(partsum)
```

Comment:
I would leave the result as is and let the calling code handle conversions. There is not much point sparsifying a fully dense result.